Исправлена ошибка

GooDiVer opened the pull request:
При переходе на форум, переведенный текст сохранялся.
При повторном открытии вкладки "Переводчик" переведенный текст не стирался

Commit:
Исправлена ошибка
При переходе на форум, переведенный текст сохранялся.
При повторном открытии вкладки "Переводчик" переведенный текст не стирался

PhBogdanov commented on app/src/main/java/e/mi/fotra/MainActivity.kt:
```diff
@@ -32,6 +35,9 @@ class MainActivity : AppCompatActivity() {
 
         main_navigation.setOnNavigationItemSelectedListener { menuItem ->
+
+            model.setTranslatedText("")
```
не очень понятно зачем тут это

PhBogdanov commented on app/src/main/java/e/mi/fotra/MainActivity.kt:
```diff
@@ -2,11 +2,14 @@ package e.mi.fotra
 class MainActivity : AppCompatActivity() {
 
     private lateinit var router: MainRouter
+    private val model: TranslateViewModel by viewModel()
```
а не sharedViewModel?

PhBogdanov replied:
я правда хз, надо читать.

PhBogdanov commented on app/src/main/java/e/mi/fotra/ViewModel/TranslateViewModel.kt:
```diff
@@ -67,6 +67,11 @@ class TranslateViewModel(
     }
 
+    fun setTranslatedText(txt: String) {
+        _translatedText.value = txt
```
таких методов во ViewModel быть не должно

во  VM должны быть только методы вроде "onBtnClicked", "onSourceTextChanged" и тд, то есть реакция на события какие-либо
а уже внутри VM может быть приватный метод setTranslatedText, если он правда нужен. 
В данном случае ты просто выставляешь дефолтное значение в активности, но это неправильно. Нужно задать это дефолтное значение именно в самой VM при ее инициализации, например. 

Иными словами, слой View не должен говорить напрямую слою VM что делать

PhBogdanov replied:
А слой View уже должен подписываться на изменения конкретных интересующих его полей. 
В общем, нарушена архитектурная задумка MVVM

Commit:
добавлены модули: ForumModule
Доавблен адаптер для отображения постов: QuestionAdapter
Добавлен интерфейс для работы с форумом: ForumService

PhBogdanov commented on app/src/main/java/e/mi/fotra/api/ForumService.kt:
```diff
@@ -0,0 +1,15 @@
+
+interface ForumService {
+
+    @Headers("Authorization: Bearer_" +
```
Так лучше не делать
Есть специальные интерцепторы, позволяющие добавлять токен 
https://proandroiddev.com/configuring-retrofit-2-client-in-android-130455eaccbd например

PhBogdanov commented on app/src/main/java/e/mi/fotra/gateway/ForumGatewayImpl.kt:
```diff
@@ -0,0 +1,29 @@
+            }
+
+            override fun onResponse(call: Call<List<Question>>, response: Response<List<Question>>) {
+                questions = response.body()
```
это не будет правильно работать. 
ты в этот объект записываешь результат уже после того как мы покинули этот метод

Есть много способов как это разрешить 
1) Прокинуть колбек в метод `getAllPost()`, 
Это может быть например такой интерфейс
nterface Callback<T> { 
      fun onSuccess(value: T)

      fun onError(e: Exception) 
}

Ты передаешь его в метод getAllPost и реализуешь где-то где у тебя есть доступ к нужным тебе данным
Это самый стандартный подход, ты точно то же самое делаешь и тут, передавая в enqueue реализацию колбека. ТОчно такое же можно сделать при вызове getAllPost()
И потом когда вернется колбек из enqueue, вызвать какой-то из методов твоего колбека (onSuccess или OnFailure)

2) Возвращать из getAllPost LiveData<List<Question>> 
Ты создаешь пустую лайфдату и возвращаешь ее из этого метода, потом где-то подписываешься на нее. Как только придет респонс, записываешь в эту лайфдату какой-то результат.

PhBogdanov replied:
Еще нужно проверять response.IsSuccessFull

PhBogdanov commented on app/src/main/java/e/mi/fotra/navigationFragments/ChooseLanguageActivity.kt:
```diff
@@ -28,11 +29,14 @@ class ChooseLanguageActivity : AppCompatActivity() {
-        })
+        adapter =
+            LanguageAdapter(
+                object :
```
не очень форматирование, раньше было лучше

PhBogdanov commented on app/src/main/java/e/mi/fotra/navigationFragments/ChooseLanguageActivity.kt:
```diff
@@ -54,7 +58,9 @@ class ChooseLanguageActivity : AppCompatActivity() {
         private val listener: Listener
-    ) : ListAdapter<Language, LanguageAdapter.LanguageViewHolder>(LanguageDiffUtilCallback) {
+    ) : ListAdapter<Language, LanguageAdapter.LanguageViewHolder>(
+        LanguageDiffUtilCallback
```
тоже особо нет смысла менять форматирование. хотя смотрится неплохо. Однако непонятно куда добавлять еще один интерфейс, если он понадобится

PhBogdanov commented on app/src/main/java/e/mi/fotra/navigationFragments/ForumFragment.kt:
```diff
@@ -5,21 +5,44 @@ import android.view.LayoutInflater
+import org.koin.androidx.viewmodel.ext.android.sharedViewModel
 
 class ForumFragment: Fragment() {
+    private lateinit var adapter: QuestionAdapter
```
зачем его делать lateinit var ? Можно инициализировать прям тут и сделать val
плюс тебе вообще не нужно иметь такое поле во фрагмнете, оно у тебя нигде не используется, кроме onActivityCreated

PhBogdanov commented on app/src/main/java/e/mi/fotra/navigationFragments/ForumFragment.kt:
```diff
@@ -5,21 +5,44 @@ import android.view.LayoutInflater
     }
 
+
     override fun onActivityCreated(savedInstanceState: Bundle?) {
```
Это неправильно, надо использовать onViewCreated

PhBogdanov commented on app/src/main/java/e/mi/fotra/navigationFragments/ForumFragment.kt:
```diff
@@ -5,21 +5,44 @@ import android.view.LayoutInflater
+
+        adapter = QuestionAdapter()
+
+        //some context
```
зачем такие комменты нужны ?)

PhBogdanov commented on app/src/main/java/e/mi/fotra/viewmodel/ForumViewModel.kt:
```diff
@@ -0,0 +1,23 @@
+    val listOfQuestions: LiveData<List<Question>>
+        get() = _listOfQuestions
+
+    fun onListOfQuestionLoaded() {
```
что это за метод вообще? Зачем он нужен?  
Ты спокойно можешь стартовать загрузку прям в init методе 

Ты его так назвал, будто это реакция на то, что у нас загрузился список вопросов, но очевидно это не так, ты его создаешь просто в момент когда у тебя создалась активити. Не имеет смысла короче

PhBogdanov commented on app/src/main/java/e/mi/fotra/gateway/ForumGatewayImpl.kt:
```diff
@@ -0,0 +1,29 @@
+            }
+
+        })
+        return questions
```
ты тут на данный момент абсолютно всегда возвращаешь null

PhBogdanov replied:
а, не совсем. Если второй раз обратиться к этому методу, в таком случае вернется предыдущее значение (которое вернулось при первом вызове)

Commit:
добавлен header interceptor. Токен задается при создании клиента

Commit:
Исправлена неправильная логика работы с колбэком.
в ForumViewModel добавлен колбэк для проброса в колбэк ретрофита
